### ISSUE TITLE
RFE: Support for reading internal cold-junction compensation temperature

### ISSUE THREAD

**Issue by HamishB** [adafruit/MAX31850_DallasTemp#3]
Hi, the Adafruit MAX31855 SPI thermocouple library has a function called readInternal() for reading the temperature off the MAX chip itself. It would be great if the library for the MAX31850 could do the same, for a couple reasons:
- hey, free data!
- I've noticed with the MAX31855 if I have a number of breakout boards together in a still box away from other heat sources etc., the internal temperature reported varies by a degree or two. By some cheap and dirty single point calibration this can be used to improve the measured thermocouple readings relative to each other, even if the thermocouples themselves are only accurate to a couple of degrees. I can only assume the MAX31850 has similar chip to chip (or is it Vreg?) variation but without a way to read the internal temperature that's unknown.
- when working near the bounds of the K-type range non-linearities creep in, and if you have both the measured and the cold-junction compensation temperatures you can work backwards to the raw voltage and apply look-up table or other mapping to more accurate temperature readings. See for example https://learn.adafruit.com/calibrating-sensors/maxim-31855-linearization

If the chip doesn't support extracting that sorry for taking your time. It wasn't clear to me from the data sheet.

many thanks,
Hamish

**Comment by ladyada** [comment]
hiya it does have the ability, its the two bytes after the compensated thermocouple temp

![image](https://user-images.githubusercontent.com/1214161/32907173-29cbac82-cacd-11e7-916f-0453299393e9.png)

but we haven't added it! not sure when we'll get to it...if you did add that capability we'd happily take a pull request :)

**Comment by HamishB** [comment]
Ok, great. In the library bytes 2 and 3 are currently mapped to the DS18B20's HIGH_ALARM_TEMP and LOW_ALARM_TEMP so something like this will be required:

```
case MAX31850MODEL:
   int16_t rawCjTemperature = (((int16_t)scratchPad[LOW_ALARM_TEMP]) << 8) | scratchPad[HIGH_ALARM_TEMP];
```

thanks,
Hamish

**Comment by ladyada** [comment]
yep! please try it out and if you can submit a pull request that would be really awesome

**Comment by HamishB** [comment]
Hi,

for your consideration the attached patch adds support for reading the internal cold-junction reference temperature from within the MAX31850 chip.

I haven't thrown it in the freezer yet to test if negative values work correctly.

example usage (precision is 0.0625 degC)
`
Serial.print("Ref Temperature for device 1 (index 0) is: ");
Serial.println(sensors.getRefTempCByIndex(0), 3);
`

regards,
Hamish
(measuring cooling rates of lava!)

[add_cold_junction_temp_MAX31855-D1wire.patch.txt](https://github.com/adafruit/MAX31850_DallasTemp/files/2295600/add_cold_junction_temp_MAX31855-D1wire.patch.txt)

**Comment by ladyada** [comment]
neat, can you try making a pull request? 😁

**Comment by SJKbiophy6** [comment]
how can i apply this patch to my library?
